### PR TITLE
Add VMSnapshot failure phase and deadline

### PR DESCRIPTION
**Pull request by ShellyKa13**
It adds a failure phase to the vm snapshot in case it takes more the defined deadline time
In such case we mark the vm snapshot as failed and clean as needed. This phase is irreversible. 

```release-note
Add phases to the vm snapshot api, specifically a failure phase
```

### CODE REVIEW

File: pkg/virt-controller/watch/snapshot/snapshot.go [hunk 1]
```diff
@@ -990,6 +1013,42 @@ func (ctrl *VMSnapshotController) checkOnlineSnapshotting(vmSnapshot *snapshotv1
...
+	log.Log.Errorf("vm Snapshot %s deadline exceeded, marking as failed", vmSnapshot.Name)
+	vmSnapshotCpy.Status.Phase = snapshotv1.Failed
+	updateSnapshotCondition(vmSnapshotCpy, newProgressingCondition(corev1.ConditionFalse, vmSnapshotDeadlineExceededError))
+	updateSnapshotCondition(vmSnapshotCpy, newFailureCondition(corev1.ConditionTrue, vmSnapshotDeadlineExceededError))
+	if !reflect.DeepEqual(vmSnapshot, vmSnapshotCpy) {
+		if _, err := ctrl.Client.VirtualMachineSnapshot(vmSnapshotCpy.Namespace).Update(context.Background(), vmSnapshotCpy, metav1.UpdateOptions{}); err != nil {
+			return err
+		}
+	}
+
```
Comment (davidvossel) [inline pkg/virt-controller/watch/snapshot/snapshot.go:1048]: I think updateSnapshotStatus should be the place that handles the status changes.

This `handleVMSnapshotDeadlineExceeded` function should understand what actions need to take place as a result of a Deadline being exceeded, and the updateSnapshotStatus should be in charge of moving the snapshot to a failed state.

**Review by davidvossel — COMMENTED** [review]
I'm seeing multiple code paths emerge here that are combining actions taken on a snapshot and updating of the snapshots status in one place. This is a difficult pattern to maintain because it makes it unclear when certain actions occur. This was somewhat less terrible before phases were introduced, but now it's starting to get more confusing. 

For example, if i delete a snapshot while in progress, will the fsUnfreeze occur?

What i'd like to see is for updateVMSnapshot to be broken into two sections. 

- a sync() which executes any actions that need to occur like fsFreeze/unFreeze. There we can have conditions like (if deletionTimestamp != nil || exceededDeadline) then unfreeze, so it becomes obvious all the conditions that should lead to an unfreeze action.
- a updateStatus() which updates the status based on what can be observed occurred as a result of a sync()

**Review by mhenriks — COMMENTED** [review]
So far so good!

Will this PR include the virt-handler changes as well?  Or will that come along later?

File: pkg/virt-controller/watch/snapshot/snapshot.go (hunk 1 above)
Comment (mhenriks) [reply pkg/virt-controller/watch/snapshot/snapshot.go:1048]: I think I understand what you are saying but I'm not sure that we are too far off from what you describe.  There is now one additional place where a snapshot is updated.  There is only one other place.  I don't think the code is hard to follow, but of course I think that

File: tests/storage/snapshot.go [hunk 2]
```diff
@@ -63,6 +66,55 @@ var _ = SIGDescribe("[Serial]VirtualMachineSnapshot Tests", func() {
...
+					AdmissionReviewVersions: []string{"v1", "v1beta1"},
+					FailurePolicy:           &fp,
+					SideEffects:             &sideEffectNone,
+					Rules: []admissionregistrationv1.RuleWithOperations{{
+						Operations: []admissionregistrationv1.OperationType{
+							admissionregistrationv1.Create,
+						},
+						Rule: admissionregistrationv1.Rule{
+							APIGroups:   []string{vsv1beta1.GroupName},
+							APIVersions: v1.ApiSupportedWebhookVersions,
```
[the remark this answers is not in the document]
Comment (ShellyKa13) [reply tests/storage/snapshot.go:99]: what do you mean? The webhook works with this config

File: pkg/virt-controller/watch/snapshot/snapshot.go [hunk 3]
```diff
@@ -185,6 +200,14 @@ func (ctrl *VMSnapshotController) updateVMSnapshotContent(content *snapshotv1.Vi
 	var volumeSnapshotStatus []snapshotv1.VolumeSnapshotStatus
 	var deletedSnapshots, skippedSnapshots []string
 
+	vmSnapshot, err := ctrl.getVMSnapshot(content)
+	if err != nil {
+		return 0, err
+	}
+	if vmSnapshotFailed(vmSnapshot) {
```
[the remark this answers is not in the document]
Comment (mhenriks) [reply pkg/virt-controller/watch/snapshot/snapshot.go:212]: Hmmm, kind of think that since the vm snapshot can never be restored (actually do we verify phase before restoring?) there is no point having volumesnapshots around.  They do take up resources on the storage backend

File: tests/storage/snapshot.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mhenriks) [reply tests/storage/snapshot.go:99]: `ApiSupportedWebhookVersions = []string{"v1alpha3", "v1"}`

Surprised that works volume snapshots which are "v1beta1"

File: pkg/virt-controller/watch/snapshot/snapshot.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ShellyKa13) [reply pkg/virt-controller/watch/snapshot/snapshot.go]: @mhenriks I just updated the code with delete of the content when deadline exceeded and although if the snapshot has failed it wont get to "Ready" state I added a cause to the restore specifically also of failed phase. last 2 commits

**Review by davidvossel — COMMENTED** [review]
in updateVMSnapshot, i'm still having difficulty following all the paths that lead to a snapshot being updated, or if double updates in a single reconcile are somehow possible, or if somehow a code path is being cut short and never executed due to an early exit/update somewhere else. 

For example, unfreezeGuestFSIFNeeded can cause a Update to occur depending of if there's an error or not, but the function calling unfreezeGuestFSIfNeeded (updateVMSnapshot) can also result in an Update being called on the snapshot.

I'd like for everything that involves performing an Update api call on a snapshot to be reduced to a single code path. 

so, 

- calculate the Snapshot status
- then a single update to write that status to the api server


This is how our core reconcile loops for vmi and migration objects work. For example look at the vmi.go controllers `updateStatus` function. We perform a bunch of calculations based on the state of things we observe in the cluster, then if the updated VMI status is different that the original VMI status (using reflect.DeepEqual), we perform a single Update call.

File: pkg/virt-controller/watch/snapshot/snapshot.go [hunk 4]
```diff
@@ -573,7 +599,12 @@ func (ctrl *VMSnapshotController) updateSnapshotStatus(vmSnapshot *snapshotv1.Vi
 		}
 	}
 
-	if vmSnapshotProgressing(vmSnapshotCpy) {
+	if vmSnapshotDeadlineExceeded(vmSnapshotCpy) {
+		vmSnapshotCpy.Status.Phase = snapshotv1.Failed
```
Comment (davidvossel) [inline pkg/virt-controller/watch/snapshot/snapshot.go:603]: while unlikely, it's logically feasible that the deadline could be reported as expired, but the content could be ready to use here. 

should this be `if !vmSnapshotReady(vmSnapshotCpy) && vmSnapshotDeadlineExceeded(vmSnapshotCpy)`

or maybe that check for ready belongs in the deadline exceeded function.

**Review by davidvossel — COMMENTED** [review]
/approve

Looks great, i just had one minor comment that doesn't necessarily even need to be addressed. just pointing a thought out.

File: pkg/virt-controller/watch/snapshot/snapshot.go (hunk 4 above)
Comment (ShellyKa13) [reply pkg/virt-controller/watch/snapshot/snapshot.go:603]: @davidvossel you are right, yet I think the idea is that even if it might have been squeezed in the timeframe once it reached the deadline the snapshot is not 100% what the user wanted since it took too long for some reason.. I think its like we draw some line which was crossed at the last minute but was crossed. so I dont mind, your call
